### PR TITLE
[RISCV] Use vwsll.vi/vx + vwaddu.wv to lower vector.interleave when Zvbb enabled.

### PR DESCRIPTION
**Pull request by yetingk**
The replacement could avoid an assignment to GPR when the type is vector of i8/i16 and vwmaccu.wv which may have higher cost than vwsll.vi/vx.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td [hunk 1]
```diff
@@ -641,6 +641,24 @@ foreach vtiToWti = AllWidenableIntVectors in {
...
                  (vti.Mask V0), GPR:$vl, vti.Log2SEW, TAIL_AGNOSTIC)>;
+
+    def : Pat<(riscv_vwsll_vl
+                 (vti.Vector vti.RegClass:$rs2),
+                 (vti.Vector (Low8BitsSplatPat (XLenVT GPR:$rs1))),
+                 (wti.Vector wti.RegClass:$merge),
+                 (vti.Mask V0), VLOpFrag),
+              (!cast<Instruction>("PseudoVWSLL_VX_"#vti.LMul.MX#"_MASK")
+                 wti.RegClass:$merge, vti.RegClass:$rs2, GPR:$rs1,
+                 (vti.Mask V0), GPR:$vl, vti.Log2SEW, TAIL_AGNOSTIC)>;
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td:652]: Should there be a _VV pattern too for completeness sake?

**Review by lukel97 — APPROVED** [review]
LGTM

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 2]
```diff
@@ -4264,24 +4264,37 @@ static SDValue getWideningInterleave(SDValue EvenV, SDValue OddV,
...
+  SDValue Interleaved;
+  if (Subtarget.hasStdExtZvbb()) {
+    // Interleaved = (OddV << VecVT.getScalarSizeInBits()) + EvenV.
+    SDValue OffsetVec =
+        DAG.getSplatVector(VecContainerVT, DL,
+                           DAG.getConstant(VecVT.getScalarSizeInBits(), DL,
+                                           Subtarget.getXLenVT()));
+    Interleaved = DAG.getNode(RISCVISD::VWSLL_VL, DL, WideContainerVT, OddV,
+                              OffsetVec, Passthru, Mask, VL);
+    Interleaved = DAG.getNode(RISCVISD::VWADD_W_VL, DL, WideContainerVT,
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:4276]: Shouldn't it be VWADDU_W_VL? We need the lower bits to be zero extended.

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**